backend/remote-state/etcdv3/backend_state.go

```diff
@@ -0,0 +1,125 @@
+package etcd
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	etcdv3 "github.com/coreos/etcd/clientv3"
+	"github.com/hashicorp/terraform/backend"
+	"github.com/hashicorp/terraform/state"
+	"github.com/hashicorp/terraform/state/remote"
+	"github.com/hashicorp/terraform/terraform"
+)
+
+const (
+	keyEnvPrefix = "-env:"
+)
+
+func (b *Backend) States() ([]string, error) {
+	client, err := b.rawClient()
+	if err != nil {
+		return nil, err
+	}
+
+	prefix := b.determineKey("")
+	res, err := client.Get(context.TODO(), prefix, etcdv3.WithPrefix(), etcdv3.WithKeysOnly())
+	if err != nil {
+		return nil, err
+	}
+
+	result := make([]string, 1, len(res.Kvs)+1)
+	result[0] = backend.DefaultStateName
+	for _, kv := range res.Kvs {
+		result = append(result, strings.TrimPrefix(string(kv.Key), prefix))
+	}
+
+	return result, nil
+}
+
+func (b *Backend) DeleteState(name string) error {
+	if name == backend.DefaultStateName || name == "" {
+		return fmt.Errorf("Can't delete default state.")
+	}
+
+	client, err := b.rawClient()
+	if err != nil {
+		return err
+	}
+
+	path := b.determineKey(name)
+
+	_, err = client.Delete(context.TODO(), path)
+	return err
+}
+
+func (b *Backend) State(name string) (state.State, error) {
+	client, err := b.rawClient()
+	if err != nil {
+		return nil, err
+	}
+
+	var stateMgr state.State = &remote.State{
+		Client: &RemoteClient{
+			Client: client,
+			DoLock: b.lock,
+			Key:    b.determineKey(name),
+		},
+	}
+
+	if !b.lock {
+		stateMgr = &state.LockDisabled{Inner: stateMgr}
+	}
+
+	lockInfo := state.NewLockInfo()
+	lockInfo.Operation = "init"
+	lockId, err := stateMgr.Lock(lockInfo)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to lock state in etcd: %s.", err)
+	}
+
+	lockUnlock := func(parent error) error {
+		if err := stateMgr.Unlock(lockId); err != nil {
+			return fmt.Errorf(strings.TrimSpace(errStateUnlock), lockId, err)
+		}
+		return parent
+	}
+
+	if err := stateMgr.RefreshState(); err != nil {
+		err = lockUnlock(err)
+		return nil, err
+	}
+
+	if v := stateMgr.State(); v == nil {
+		if err := stateMgr.WriteState(terraform.NewState()); err != nil {
+			err = lockUnlock(err)
+			return nil, err
+		}
+		if err := stateMgr.PersistState(); err != nil {
+			err = lockUnlock(err)
+			return nil, err
+		}
+	}
+
+	if err := lockUnlock(nil); err != nil {
+		return nil, err
+	}
+
+	return stateMgr, nil
+}
+
+func (b *Backend) determineKey(name string) string {
```

Comment:
If the changes in etcd means that there isn't going to be a direct upgrade from etcd2, you can probably get rid of the legacy default state path, and assume everything has a "workspace" and get rid of the conditionals around `keyEnvPrefix`and such.

Reply:
> (...) you can probably get rid of the legacy default state path (...) and get rid of the conditionals around `keyEnvPrefix` and such.

@jbardin I'm not sure I understand what you mean. 😶 Could you please clarify/exemplify?

Reply:
I meant that the reason the other backends have the "default" state handled separately is solely for backwards compatibility with state files that existing before envs/workspaces.

So rather than dealing with the `keyEnvPrefix` you've adopted from the others, you could use the same hierarchy in all cases of `prefix/name`, and just include "default" in there.

Reply:
Done @ 3c21b9c. Thank you for your guidance. 🙂